GRID-384: Plugins

joneit opened the pull request:
Got inspired on the train. Hope y'all like it.

joneit commented on demo/js/tree-view.js:
```diff
@@ -22,21 +17,15 @@ window.onload = function() {
+        Hypergrid.rowById, // ditto
+        Hypergrid.Hyperfilter, // object API instantiation; `$$CLASS_NAME` defined so ref saved in `grid.plugin.hyperfilter`
+        Hypergrid.Hypersorter, // object API instantiation to grid.plugin; no `name` or `$$CLASS_NAME` defined so no ref saved
+        ['treeviewAPI', Hypergrid.TreeView, options] // object API instantiation with one arg; [0] overrides any defined name so ref saved in `grid.plugin.treeViewAPI`
```
Maybe an object here? I usually use objects. The only reason I went with an array here is it "mimics" the constructor args. Alternative would look like:
```javascript
        { name: 'treeviewAPI', API: Hypergrid.TreeView, options, args: options }
```

Dwaynekj replied:
I'd vote for object.
An object to me signifies all of the properties as having a relationship. An Array not so much

joneit commented:
@Dwaynekj Please finish reviewing. Thanks.

Dwaynekj commented:
@joneit Working on sorting. I don't think we need to prioritize this for review (as it will take time and not originally apart of the sprint)

Dwaynekj commented on src/Hypergrid.js:
```diff
@@ -263,6 +266,82 @@ Hypergrid.prototype = {
+     * The resulting plain object or instantiated objects may be named by (in priority order):
+     * 1. if `pluginSpec` contains an array and first element is a string
+     * 2. object has a `name` property
+     * 3. object has a `$$CLASS_NAME` property
```
Should consider if extend-me specific properties will cause issues if we ever switch to ES6

Dwaynekj replied:
More importantly. We should also support an uninstall plugin

Dwaynekj replied:
Also. We should have warnings if a plugin collides with an existing property on one of the Core objects. As in someone wrote a method and mistakenly called it setData

Dwaynekj commented on src/Hypergrid.js:
```diff
@@ -67,7 +67,7 @@ function Hypergrid(container, options) {
         container = null;
     }
-    this.options = options || {};
+    this.options = options = options || {};
```
huh?

Dwaynekj commented on src/Hypergrid.js:
```diff
@@ -78,7 +78,7 @@ function Hypergrid(container, options) {
     this.numColumns = 0;
     this.clearMouseDown();
-    this.setFormatter(this.options.localization);
+    this.setFormatter(options.localization);
```
I don't see the need of getting rid of the this keyword

Dwaynekj commented on demo/js/tree-view-separate-drill-down.js:
```diff
@@ -22,21 +18,15 @@ window.onload = function() {
+    grid = new Hypergrid('div#tree-example', { data: treeData, plugins: [
+        Hypergrid.drillDown, // simple API install (plain object with `install` method) but no `name` defined so no ref is saved
+        Hypergrid.rowById, // ditto
+        Hypergrid.Hyperfilter, // object API instantiation; `$$CLASS_NAME` defined so ref saved in `grid.plugin.hyperfilter`
```
Definitely looks cleaner

Dwaynekj commented on src/Hypergrid.js:
```diff
@@ -263,6 +266,82 @@ Hypergrid.prototype = {
+     *    * remaining arguments are optional arguments for the object constructor
+     * * falsy value such as `undefined` - ignored
+     *
+     * The API may have a `name` or `$$CLSSS_NAME` property.
```
Spelling Error $$CLSSS_NAME

Dwaynekj commented:
@rdepena I think this one is important for you to look at as well.
@joneit has a plugins implementation, but we don't really have a product definition on what a Hypergrid plugin `is` and `isn't`. Or is it anything?

rdepena commented:
@Dwaynekj @joneit I kinda agree with DJ, can we take a step back instead of reviewing the code, can we have a high level conversation on what plugins are and how they integrate into hypergrid.

Dwaynekj commented:
@rdepena @joneit Discussed in person. We will expose this architecture with the caveat that its exploratory and it is subject to changed. We must the user well aware of this.

I do feel it may be too powerful and would be hard to rescind the idea later. If we had a formal and properly pruned and public API (long time coming now) then we would expose hooks that plugins would be allowed to use.

At the moment we are saying a plugin can do anything it wants.

👍

joneit commented:
* Changed name of plugins hash from `grid.plugin` to `grid.plugins` _**throughout**_
* Added `unintallPlugins` method, which can be called like this:
```javascript
grid.uninstallPlugins(); //uninstalls all
grid.uninstallPlugins('myPlugin'); //uninstalls just this one
grid.uninstallPlugins(['myPlugin', 'myOtherPlugin']);
```

Dwaynekj commented:
🎉
